platform/linux-generic/odp_crypto.c

```diff
@@ -308,6 +317,148 @@ void packet_hmac(odp_packet_t pkt,
 	HMAC_Final(ctx, hash, NULL);
 }
 
+static void do_pad_xor(uint8_t *out, const uint8_t *in, int len)
+{
+	int pos = 0;
+
+	for (pos = 1; pos <= 16; pos++, in++, out++) {
+		if (pos <= len)
+			*out ^= *in;
+		if (pos > len) {
+			*out ^= 0x80;
+			break;
+		}
+	}
+}
+
+static void xor_block(uint32_t *res, const uint32_t *op)
+{
+	res[0] ^= op[0];
+	res[1] ^= op[1];
+	res[2] ^= op[2];
+	res[3] ^= op[3];
+}
+
+static
+void packet_aes_xcbc_mac(odp_packet_t pkt,
+		 const odp_crypto_packet_op_param_t *param,
+		 odp_crypto_generic_session_t *session,
+		 uint8_t *hash)
+{
+	uint32_t e[4] = {0, 0, 0, 0};
+	uint32_t offset = param->auth_range.offset;
+	uint32_t len   = param->auth_range.length;
+	uint32_t seglen = 0;;
+	int dummy_len = 0;
+	EVP_CIPHER_CTX *ctx;
+	void *mapaddr;
+	uint8_t *data;
+	
+	ODP_ASSERT(offset + len <= odp_packet_len(pkt));
+
+	ctx = EVP_CIPHER_CTX_new();
+	EVP_EncryptInit_ex(ctx, session->auth.evp_cipher, NULL, session->auth.key, NULL);
+	mapaddr = odp_packet_offset(pkt, offset, &seglen, NULL);
```

Comment:
This is actually even worse than calling `odp_packet_data()`. You still access all package data as a contiguous memory area.
Note to @psavol : even updated documentation doesn't really help people from making such mistake
Note to @lumag : we need to update odp_crypto tests to include test on segmented packets

Reply:
so how to make it work for segmented packets?

Reply:
take a look at _packet_compare_data() in api tests how it compares data of 2 packets in ./test/validation/api/packet/packet.c

Reply:
@SonicwallYhe 
Or you can take a look onto packet_hmac. Basically you have to go through the full packet, generating MAC. At each moment you either have full AES_BLOCK_SIZE contiguous memory, or you have to copy contents of the packet to interim memory location, use it to as the next iteration data and then continue from next offset.